[댓글 기능] 에헴 미션 제출합니다.

hyogenie-v opened the pull request:
안녕하세요 에헴입니다.
짧은 기한 내에 미션을 진행하느라 부족한 점이 많습니다.

피드백 부탁드립니다. 감사합니다~

ike5923 committed:
[게시글 생성/조회/수정/삭제] 아이크 미션 제출합니다. (#18)

* feat: add gradle dependency

* feat: get/post practice

* refactor: modify path

* fix: add responseBody annotation

* feat: click the button to move article-edit page

* test: modify uri

* feat: create and save article

* feat: write the article

* feat: look up the article

* feat: look up the article by id

* fix: find an article by id

* feat: update the article

* feat: delete the article

* feat: move to index from article-edit

* feat: setting devtools

* fix : UI bug

* refactor: remove duplicated html code

* refactor: remove duplicated index logo

* refactor: create ArticleService

* refactor: divide package

* refactor: divide package

* refactor: modify return value

* refactor: modify test cod

- fix bug
- extract duplicate fields

* feat: update README.md

* refactor: modify variable naming

* refactor : remove Article set method

* refactor: change from getting id to sending message to article

* refactor: ensure the list does not change when it is passed

* refactor: check convention

* refactor: check convention

* refactor: change path (absolute -> relative)

* refactor: move devtools property

* test: article update

* test: articleRepository save

* test: articleRepository findById

* test: articleRepository delete

* refactor: remove ArticleService

* refactor: test code convention

ike5923 committed:
[게시판 회원 기능] 아이크 미션 제출합니다 (#50)

* build: jpa 추가

* feat: ArticleRepository 생성

1. JPA CrudRepository 상속.
2. 학습테스트

* feat: MySQL 연동

1. application.properties 설정
2. 마크업 정리

* test: ArticleControllerTest

1. ArticleRepository 이용하도록 변경
2. URL 수정

* feat: gitignore

bin/ 추가

* feat: User Entity 생성

1. UserRepository 준비
2. UserController 생성
3. user-list.html 마크업 수정

* feat: UserController 생성

1. 엔드포인트 생성
2. 마크업 변경

* feat: 로그인 / 회원가입 구현

* feat: User 유효성 검사 추가

1. UserException 생성
2. 유저 이름, 비밀번호 정규식 매칭 로직 추가

* refactor: User.isEqualTo() 메서드 생성

1. 유저가 입력한 로그인 정보와 매칭하는 메서드
2. equals hashcode

* refactor: UserController 메서드 변경

1. Post /login /logout
2. HttpSession 사용

* refactor: 템플릿 변경

1. signup.js : 두 번 입력한 패스워드가 서로 다를 경우 submit하지 않음.
2. 로그인 시 드롭다운 제어.

* setting: 설정 변경

1. hibernate show sql

* test: UserRepositoryTest 생성

* feat: UserService 추가

1. 서비스와 컨트롤러간 데이터 이동을 위한 Dto 생성

* feat: LoginInterceptor 생성

1. LoginInterceptorConfig 생성
2. 인터셉터 제외 URL 추가.

* feat: ArticleService 생성

1. 서비스 레이어 추가
2. ArticleControllerTest 기본 인증 추가

* feat: UserRepository 생성

1. findByEmail 메서드 추가

* refactor: navbar 템플릿 조정

1. dropdown.html 삭제

* feat: User 수정, 탈퇴 구현

1. Put, Delete 매핑
2. UserService 기능 추가
3. 테스트 작성

* test: LoginInterceptorConfigTest 테스트

1. 인터셉터 테스트

* test: 모든 테스트 확인

1. ddl-auto : create 로 변경
2. BeforeEach, AfterEach 추가

* setting: 로그 설정

1. 테스트 패키지와 프로덕트 패키지 설정 분리
2. 템플릿 중복 제거

* docs: README.md 추가

* setting: db 설정 변경

1. mysql version 적용
2. db 연결 정보 변경

* refactor: iterable to list - stream 적용

* refactor: convetion 정리

* refactor: renaming

-
1. articleDto -> articleRequestDto
2. userDto -> userRequestDto

* refactor: correct the typing error

* refactor: change password validate pattern

1. at least one uppercase character
2. at least one lowercase character
3. at least one special character
4. at least one number

* refactor: naming clearly

* feat: create LoginService

* refactor: make session info class

* refactor: remove duplicate test code

1. 회원가입 template 생성
2. 회원가입을 상속받아 로그인 template 생성
3. 로그인 template을 상속받아 controller test 중복 제거

* feat: use logger

model과 session의 log를 확인하기 위해
관련 method를 util로 분리 후 logger사용

* feat: apply ControllerAdvice

create controllerExceptionHandler

* refactor: change mapping path naming

* feat: exception handling using ControllerAdvice

remove loginDto, SignUpDto

* feat: page access control(login, signup)

login 되어있을 때, login, signup 접근 불가

* refactor: Not know session at service layer

* refactor: session DI

* feat: apply interceptor - logged in access page

* refactor: remove not using method

* refactor: send exception message

* refactor: unable to reallocate userRepository

* feat: manage transaction

* feat: use saved result

* feat: add entity column

* feat: apply @Valid at userRequestDto

* feat: change session data (user info -> user)

* feat: apply converter and remove toEntity

* feat: apply pagination

* refactor: page info constant

* refactor: remove unnecessary converter config

* refactor: change log level at throw exception

* feat: create PageRequest class and validate page request data

* refactor: arrnage package

* refactor: unnecessary rollback option

* refactor: change converter (spring interface method to static method)

hyogenie-v committed:
feat: 뷰 세션에서 User 객체가 아닌 UserResponseDto 객체를 갖도록 수정

1. UserResponseDto 클래스를 추가한다
2. UserController 클래스에서 세션에 User 객체 대신 UserResponseDto 객체를 사용하도록 수정한다

hyogenie-v committed:
test: UserService에 대한 테스트 클래스 및 메서드 추가

hyogenie-v committed:
feat: 로그인 여부와 상관없이 게시글 상세 조회 페이지에 접속할 수 있도록 변경

hyogenie-v committed:
feat: User 객체와 Article 객체의 관계 매핑

hyogenie-v committed:
feat: 게시글을 작성하였을 때 글쓴이의 정보도 함께 저장하는 로직 추가

hyogenie-v committed:
refactor: Repository 인터페이스가 JpaRepository 인터페이스를 상속하도록 수정

hyogenie-v committed:
feat: Comment 엔티티 클래스 생성 및 연관 관계 매핑

1. Comment 엔티티 클래스를 추가한다
2. Article, User, Comment의 연관 관계를 매핑한다
3. CommentRepository 클래스를 추가하고 findByArticle() 메서드를 추가한다
4. 테스트 클래스를 추가하고 특정 게시글에 대한 모든 댓글을 찾아오는지 테스트한다

hyogenie-v committed:
feat: 댓글 작성 및 조회 기능 추가

1. 댓글에 대한 기능을 관장하는 CommentService 클래스를 추가한다
2. 게시글 삭제 시 해당 게시글에 속해있는 모든 댓글을 삭제하도록 기능을 추가한다
3. 게시글에 속해있는 모든 댓글들을 가져오는 메서드를 추가한다
4. model에 댓글 데이터를 추가한다
5. article_comment 뷰에 댓글을 저장하는 버튼을 추가한다
6. 전달받은 comments 데이터를 뷰에 출력하는 기능을 추가한다

hyogenie-v committed:
feat: 댓글 수정 및 삭제에 대한 비즈니스 로직 구현

1. CommentController 클래스에 댓글을 수정 및 삭제하는 메서드를 추가한다
2. CommentService 클래스에서 댓글을 수정 또는 삭제하려는 사용자가 댓글 작성자인지 확인한다
3. CommentException 클래스와 CommentAuthenticationException 클래스를 추가한다
4. ExceptionHandler 클래스에 댓글과 관련된 예외에 대한 처리를 추가한다

hyogenie-v committed:
feat: 뷰에 댓글을 수정하고 삭제하는 기능을 추가한다

1. 뷰에 댓글을 수정 또는 삭제하는 기능을 추가한다
2. 댓글을 수정 또는 삭제하려는 사용자가 댓글 작성자인지 검증한다

hyogenie-v committed:
refactor: code convention refactoring

hyogenie-v committed:
revert: 프로젝트 초기화

phs1116 reviewed (changes requested):
안녕하세요. 리뷰어 화투입니다. :)
짧은시간 내에 정말 구현 잘 하셨어요. 👍 
다만 Comment에 대한 인수테스트 및 예외케이스,
몇가지 연관관계를 활용하면 좋을것 같은점들 피드백 추가했습니다.
확인 부탁드리고 궁금하신 점 있으시면 연락주세요. :)

hyogenie-v committed:
fix: Entity 객체의 동등성 검사 기준을 id로만 갖도록 수정

hyogenie-v committed:
refactor: Converter 클래스를 엔티티 별로 분리

hyogenie-v committed:
refactor: Article 객체가 author 멤버의 setter를 사용하지 않도록 수정

hyogenie-v committed:
test: CommentController 클래스에 대한 테스트 메서드 추가

1. CommentController 클래스에 대한 테스트 메서드를 추가한다
2. 임의의 User를 이용한 테스트를 위해 테스트 템플릿 메서드를 오버로딩한다

hyogenie-v committed:
refactor: 예외 클래스가 메시지를 갖도록 수정 및 예외 상속 클래스 추가

hyogenie-v committed:
refactor: 패키지 구조를 기능별로 분리

hyogenie-v committed:
feat: 게시글 작성자가 아닌 사용자가 글을 수정하거나 삭제하려는 경우에 대한 예외처리 추가

1. ArticleService 클래스에 유효하지 않은 수정 또는 삭제의 경우에 예외를 던지는 메서드를 추가한다
2. ArticleService 클래스의 checkAuthentication 메서드에 대한 예외 테스트를 추가한다
3. ArticleController 클래스에 checkAuthentication 메서드에 대한 예외 테스트를 추가한다

hyogenie-v committed:
test: Controller 테스트에서 repository 빈 객체를 사용하지 않고 api를 호출하여 테스트하도록 수정

hyogenie-v committed:
feat: Article과 Comment 객체 관계를 양방향으로 매핑하도록 수정

hyogenie-v committed:
test: Repository 테스트 메서드 추가

1. User 엔티티 객체의 구조에 맞게 테스트를 수정한다
2. Article 객체와 User 객체의 연관관계에 대한 저장 테스트를 추가한다
3. Article 객체와 Comment 객체의 연관관계에 대한 저장 및 삭제 테스트를 추가한다
4. RepositoryTest에서 공통적으로 사용하는 상수를 분리한 StaticVariableTemplate 클래스를 추가한다

hyogenie-v committed:
feat: User 및 Article 객체의 update 메서드에 권한 체크 로직 추가

1. User 및 Article 객체에서 데이터 업데이트가 발생할 때 권한을 체크하도록 로직을 추가한다
2. Service 클래스의 update 메서드에 유저 정보에 대한 파라미터 변수를 추가한다
3. 각 Service 클래스가 자신 외의 Repository 객체를 갖지 않고 Serivce 객체를 갖도록 수정한다

hyogenie-v committed:
test: Article 클래스의 수정으로 인한 update 메서드 테스트 수정

hyogenie-v commented:
안녕하세요 화투님!
피드백을 반영하면서 참고해야 하는 것들이 많아 리뷰 요청이 예상보다도 늦은 것 같습니다...!

아직 DTO의 사용이 익숙치 않고 사용할 만한 경우를 구분하는 것이 어려워서 제가 봐도 얼룩덜룩한 부분들이 많습니다.
주신 피드백에 대해 나름대로의 생각을 적긴 했는데 크루들과 이야기를 나눌 때마다 생각이 달라져서 느낀 바를 토대로 다시 수정을 해야 할 것 같습니다ㅜㅜ
제가 달은 댓글에 대해 어떻게 생각하시는지 의견을 달아주시면 수정할 때 많은 도움이 될 것 같아요! 👀 

JPA는 처음 사용해보는 거라 단방향과 양방향이 사용하기 적절한 경우에 대한 기준이 명확하지 않아서 엔티티간 연관관계를 지정해주는 게 많이 헷갈리고 어려웠습니다.

피드백 주신대로 `Article`과 `Comment`의 관계를 양방향으로 변경해보았는데 fk를 갖고 있지 않은 자식 엔티티인 `Article` 객체에서 댓글을 탐색하고 추가하는 게 훨씬 간편하다고 느꼈습니다.
하지만 테스트를 작성하다보니 부딪힌 어려움이 있어서 추가적으로 질문을 드리고 싶은데요,

**게시글이 삭제되면 댓글도 함께 삭제되는지에 대한 테스트**를 진행할 때
1. `ArticleRepositoryTest`에서 확인하려고 해도 해당 게시글이 지워졌기 때문에 댓글 삭제 여부를 확인하려면 `CommentRepository`가 필요하고
2. 똑같이 `CommentRepositoryTest`에서도 게시글을 삭제하기 위해 `ArticleRepository`를 호출해야 했습니다.

편리한 테스트를 위해서 다른 빈 객체를 가져다 사용하는 건 한 빈 객체에 대한 순수한(?) 테스트가 아닌 것 같아서 곤란했습니다.

그래서 **`CommentRepositoryTest`에서 Persistence Context와 동기화하기 전에 `testEntityManager`의 `remove()` 메서드를 사용해서 Article 객체를 삭제**하도록 짜기는 했는데 적절한 방법이었는지 궁금합니다.
리뷰와 함께 이런 경우에는 어떻게 해야 할 지 추가적인 방향이나 힌트를 제시해주셨으면 합니다.

감사합니다!

hyogenie-v committed:
fix: 순환참조 문제 수정 및 연관관계 매핑 설정 수정

1. ArticleService에서 CommentService 필드를 삭제한다
2. Comment 엔티티에서 User 매핑에 대한 fetch 유형을 EAGER로 수정한다

phs1116 reviewed (changes requested):
안녕하세요! 리뷰어 화투입니다.
전체적으로 매우 잘 작성해주셨어요. 개선 👍 
질문주신 사항 및 몇가지 피드백 추가했습니다.
참고 부탁드리구요 궁금하신 점 있으시면 DM 주세요 :)

hyogenie-v committed:
refactor: checkAuthentication 메서드의 조건절이 메시지를 통해 권한을 확인하도록 수정

1. System.out.println 를 삭제하고 로그를 사용하도록 수정한다
2. checkAuthentication 메서드의 조건절이 메시지를 통해 권한을 확인하도록 수정한다
3. Converter 클래스의 메서드명을 역할이 드러나도록 수정한다
4. 프로덕션 코드의 수정사안을 테스트 코드에 반영한다

hyogenie-v committed:
feat: 게시글 및 댓글 권한 검사를 서비스에서 담당하도록 수정

hyogenie-v committed:
feat: ArgumentResolver를 이용해 세션에서 사용자 데이터를 가져오는 중복된 코드 개선

1. LoginUserArgumentResolver 클래스를 추가하여 사용자 정보를 처리하는 로직을 추가한다
2. 설정 정보에 추가한 ArgumentResolver를 등록한다
3. Article, Comment, User 컨트롤러의 메서드에 로그인된 사용자의 객체를 파라미터로 받도록 수정한다

phs1116 reviewed (approved):
안녕하세요! 리뷰어 화투입니다.
구현 매우 잘 해주셨습니다 👍 
이번 단계 머지하도록 하겠습니다.
개발 하시느라 수고 많으셨어요 :)
